Fix: HPA suppresses FailedRescale event on successful conflict retry.

AumPatel1 opened the pull request:
This change modifies the HPA controller to only emit a FailedRescale event if retrying the scale operation (due to a conflict) ultimately fails. If the retry succeeds, a SuccessfulRescale event is emitted, reducing noise from transient conflicts. Addresses #132002

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug


#### What this PR does / why we need it:
This PR addresses issue #132002, where the Horizontal Pod Autoscaler (HPA) reports a FailedRescale error due to a race condition. The error occurs when the HPA controller attempts to update the target scale, but the operation fails because the scale object has been modified by another process.

Changes Made:
Implemented retry logic using retry.RetryOnConflict to handle transient conflicts gracefully.
Removed redundant assignment to ensure the code uses the latest version of the scale object.
Updated error handling to set the appropriate condition based on whether the error is a conflict or not.
Expected Behavior:
The HPA controller will now retry the scale update operation if a conflict is detected, reducing the frequency of FailedRescale errors.
This change improves the reliability of the HPA controller and reduces alert noise.

#### Which issue(s) this PR fixes:
Fixes #132002


#### Special notes for your reviewer:
Testing:
Unit tests have been added to verify the retry logic and error handling.
All tests are passing.


#### Does this PR introduce a user-facing change?
```release-note
HPA controller will no longer emit a 'FailedRescale' event if a scale operation initially fails due to a conflict but succeeds after a retry; a 'SuccessfulRescale' event will be emitted instead. A 'FailedRescale' event is still emitted if retries are exhausted.
```

pacoxu commented:
/ok-to-test

raywainman commented:
/retest

AumPatel1 commented:
Hi @mwielgus and @raywainman, I've addressed the issue with the HPA controller as per #132002. The PR implements retry logic to handle conflicts gracefully, and I've added unit tests to verify the behavior. All checks have passed, and I've force-pushed the latest changes. Could you please take a look ,Thanks!

omerap12 reviewed (commented):
Thanks for this! I left some comment :)

omerap12 commented:
/triage-accepted

omerap12 commented:
/triage accepted

omerap12 commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -867,17 +868,57 @@ func (a *HorizontalController) reconcileAutoscaler(ctx context.Context, hpaShare
+		})
+
 		if err != nil {
+			// Add logging as requested by the reviewer
+			logger := klog.FromContext(ctx)
+			logger.Error(err, "Failed to scale target after retries", "HPA", klog.KObj(hpa), "scaleTarget", reference)
```
```diff
                if err != nil {
-                       // Add logging as requested by the reviewer
-                       logger := klog.FromContext(ctx)
-                       logger.Error(err, "Failed to scale target after retries", "HPA", klog.KObj(hpa), "scaleTarget", reference)
+                       logger.Error(err, "Failed to scale target after retries",
+                               "HPA", klog.KObj(hpa),
+                               "scaleTarget", reference,
+                               "desiredReplicas", desiredReplicas,
+                               "currentReplicas", currentReplicas,
+                               "error", err.Error())
```

AumPatel1 replied:
i changed it , is it good . and also  i tried removing "%s" and passing ableToScaleMessage directly, but the Go compiler errored with "non-constant format string in call to setCondition " because setCondition uses fmt.Sprintf(message, args...) internally, and Go requires message to be a constant string literal. Using "%s", ableToScaleMessage ensures a constant format string while correctly passing the message.

omerap12 reviewed (changes requested):
nit:

omerap12 commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -867,17 +868,62 @@ func (a *HorizontalController) reconcileAutoscaler(ctx context.Context, hpaShare
+		})
+
 		if err != nil {
+			// Add logging as requested by the reviewer
```
Please remove this line

omerap12 commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -867,17 +868,62 @@ func (a *HorizontalController) reconcileAutoscaler(ctx context.Context, hpaShare
+
 		if err != nil {
+			// Add logging as requested by the reviewer
+			logger := klog.FromContext(ctx)
```
Do we need this? logger is already initiated here: https://github.com/kubernetes/kubernetes/pull/132007/files#diff-f1c5a31aa8fb8e3fd64b6aa13d3358b504e6e25030f249f1652e244c105eafc7R813

omerap12 reviewed (commented):
Thanks! 
/lgtm

soltysh commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -867,17 +868,60 @@ func (a *HorizontalController) reconcileAutoscaler(ctx context.Context, hpaShare
+			latestScale.Spec.Replicas = desiredReplicas
+			_, err = a.scaleNamespacer.Scales(hpa.Namespace).Update(ctx, targetGR, latestScale, metav1.UpdateOptions{})
+			return err
+		})
```
In the linked issue you're mentioning that "Occasionally, we observe the following error during scaling...", but the current implementation changes the order such that currently we'll do:
1. get the scale
2. decide if we need to scale
3. update scale
whereas previously we'd only invoke a single update scale operation. Iow. from previous single operation (just update) you're introducing 2 explicit invocations (get + update).

The better approach will be to invert the logic like so:
1. try to update
2. if no error - return
3. if error - get the scale, compare replicas 
4. make sure to always return the error from the update operation (step 1), so that you'll trigger the retry.

This code in statefulset has this nicely laid out:
https://github.com/kubernetes/kubernetes/blob/08ee8bde594a42bc1a222c9fd25726352a1e6049/pkg/controller/statefulset/stateful_set_status_updater.go#L52-L73

This approach will give you the retry benefit you're looking for, but at the same time ensuring that in case of a successful first update you won't invoke more than a single update operation.

soltysh commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -867,17 +868,60 @@ func (a *HorizontalController) reconcileAutoscaler(ctx context.Context, hpaShare
+			// Reason indicates *why* this scaling attempt failed in a way that overrides min/max considerations.
+			setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionFalse, "ScaleUpdateFailed", "The HPA controller was unable to update the target scale.")
+
+			// Record the event indicating the ultimate failure
```
I don't see a reason we should be changing the events nor conditions here. The current ones are sufficient and they are explaining the situation sufficiently, since they already include the details of the error inside. Why do we need additionally state that we retried? What it gives us? I'd suggest to go with just the simplest change for starters, which is adding the retries as pointed above.

soltysh reviewed (commented):
/lgtm cancel
see my comments about simplifying this approach

AumPatel1 commented:
Hello @soltysh and @omerap12,
Thank you for all your detailed feedback and patience. I have now implemented the retry logic using retry.RetryOnConflict as you suggested.

omerap12 commented:
Can you please fix the commit message? 
"[Keywords](https://help.github.com/articles/closing-issues-using-keywords) which can automatically close issues and at(@) or hashtag(#) mentions are not allowed in commit messages."

AumPatel1 commented:
/retest

soltysh reviewed (approved):
/approve

there's one more minor nit and that e2e failure which needs to be resolved

@omerap12 has the final lgtm on that PR

AumPatel1 committed:
Fix: HPA suppresses FailedRescale event on successful conflict retry

This change modifies the HPA controller to use retry.RetryOnConflict when updating a scale subresource. This prevents the controller from emitting a FailedRescale event on transient API conflicts if a subsequent retry succeeds. If the retry is successful, a SuccessfulRescale event is emitted. If all retries are exhausted and the conflict persists, the original FailedRescale event is emitted. This reduces event noise caused by race conditions where the scale subresource is updated by another process.

omerap12 reviewed (approved):
Thanks for your contribution! 
/lgtm

omerap12 commented:
/lgtm cancel

omerap12 reviewed (commented):
on second thought, can we add a unit test for this change?

omerap12 reviewed (approved):
Thanks!
/lgtm
